src/core/geometry/qgsgeometry.cpp

```diff
@@ -994,6 +995,81 @@ QgsGeometry QgsGeometry::orientedMinimumBoundingBox( double &area, double &angle
   return minBounds;
 }
 
+static QgsCircle __recMinimalEnclosingCircle( QgsMultiPoint points, QgsMultiPoint boundary )
+{
+  auto l_boundary = boundary.length();
+  QgsCircle circ_mec;
+  if ( ( points.length() == 0 ) || ( l_boundary == 3 ) )
+  {
+    switch ( l_boundary )
+    {
+      case 0:
+        circ_mec = QgsCircle();
+        break;
+      case 1:
+        circ_mec = QgsCircle( QgsPoint( boundary.last() ), 0 );
+        boundary.pop_back();
+        break;
+      case 2:
+      {
+        QgsPointXY p1 = boundary.last();
+        boundary.pop_back();
+        QgsPointXY p2 = boundary.last();
+        boundary.pop_back();
+        circ_mec = QgsCircle().from2Points( QgsPoint( p1 ), QgsPoint( p2 ) );
+      }
+      break;
+      default:
+        QgsPoint p1( boundary.at( 0 ) );
+        QgsPoint p2( boundary.at( 1 ) );
+        QgsPoint p3( boundary.at( 2 ) );
+        circ_mec = QgsCircle().minimalCircleFrom3points( p1, p2, p3 );
+        break;
+    }
+    return circ_mec;
+  }
+  else
+  {
+    QgsPointXY pxy = points.last();
+    points.pop_back();
+    circ_mec = __recMinimalEnclosingCircle( points, boundary );
+    QgsPoint p( pxy );
+    if ( !circ_mec.contains( p ) )
+    {
+      boundary.append( pxy );
+      circ_mec = __recMinimalEnclosingCircle( points, boundary );
+    }
+  }
+  return circ_mec;
+}
+
+QgsGeometry QgsGeometry::minimalEnclosingCircle( QgsPoint &center, double &radius, unsigned int segments ) const
+{
+  center = QgsPoint( );
+  radius = 0;
+
+  if ( !d->geometry )
+  {
+    return QgsGeometry();
+  }
+
+  /* optimization */
+  QgsGeometry hull = convexHull();
+  if ( hull.isNull() )
+    return QgsGeometry();
+
+  QgsMultiPoint P = hull.convertToPoint( true ).asMultiPoint();
```

Comment:
For reference: This part could be optimised considerably when https://github.com/qgis/QGIS/pull/4513 lands. But it's all private anyway and none of this is locked into the API, so we can address this after the iterator work is finished.

Reply:
Nice!
I take this opportunity to specify that I used the convex envelope as the first filter because otherwise (on my VM) with the recursion QGis can crash with a significant number of points (~ 10 000).